[tools][docs] bump TypeDoc, migrate to @example annotation

Simek opened the pull request:
# Why & how

This PR updates TypeDoc to the latest version which fixes the `isConst` flag issue and introduces again some performance gains (reduced files generation time).

Unfortunately one of the breaking changes in new release was a change of Projects and Enums `kind` value, which resulted in regeneration of all doc data files (https://github.com/TypeStrong/typedoc/releases/tag/v0.22.0):
> The ReflectionKind values for `Project`, `Module`, `Namespace`, and `Enum` have changed.

The backwards compatibility has been retained, because fortunately new Enum value was not used before. It only required small change to the `filterDataByKind`, which now can also accept arrays of kind types. One of the tests cases already run on SDK 42 file, so there was no need to introduce new test case to cover this breaking change.

I have decided to leverage this situation and I have also make a refactor of all code example in comments to use `@example` annotations, which enables more unified styling, fixes the issue when in certain situations example has been interpreted as `@param` description and it is standard-correct approach.

Also few more tweaks to the `APISection` components and utils has been applied:
* added `getTagData` helper function to fetch annotation tags data
* simplified code block language detection in Markdown renderer
* small tweaks to Example section rendering
* few redundant whitespace cleanups in generated code (example changes visible in test snapshot diff)

# Test Plan

Docs API data generation command works correctly after TypeDoc update, generated JSONs and component changes have been verified by running docs website on `localhost`. 

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

# Preview
###  Example code block before and after the change
<img align="left" width="380" alt="Screenshot 2021-09-18 at 14 39 29" src="https://user-images.githubusercontent.com/719641/133889105-2eef667b-5b52-4f99-9b97-cc1d847ac5db.png">
<img width="380" alt="Screenshot 2021-09-18 at 14 39 04" src="https://user-images.githubusercontent.com/719641/133889106-7a6dbcd9-1fea-47be-9c74-8983061aee5f.png">

Simek committed:
[tools][docs] bump TypeDoc, migrate to @example annotation

Simek committed:
small fix in Markdown renderer

Simek committed:
lint fixes

ide commented on docs/components/plugins/APISection.tsx:
```diff
@@ -22,12 +22,14 @@ type Props = {
 
 const filterDataByKind = (
   entries: GeneratedData[],
```
```suggestion
  entries: GeneratedData[] = [],
```

ide commented on docs/components/plugins/api/APISectionUtils.tsx:
```diff
@@ -37,11 +38,9 @@ export const mdComponents: MDComponents = {
-      </Code>
+  code: ({ children, className }) =>
+    className ? (
+      <Code className={className || 'language-unknown'}>{children}</Code>
```
If className is truthy here, the prop's value expression will never evaluate to `language-unknown`

Simek replied:
Yup, it is leftover from previous implementation, I clean this up! 🙂

ide commented on docs/components/plugins/api/APISectionUtils.tsx:
```diff
@@ -306,6 +305,9 @@ export const getCommentOrSignatureComment = (
   signatures?: MethodSignatureData[]
 ) => comment || (signatures && signatures[0]?.comment);
 
+export const getTagData = (tagName: string, comment?: CommentData) =>
+  comment?.tags?.filter(tag => tag.tag === tagName)[0];
```
This function seems useful only if CommentData is passed in -- make it the caller's responsibility not to call this function if the CommentData is falsy.

Simek replied:
I have tried refactor this method to the recommended approach, but it would result in type incompatibility, since everywhere else in types `CommentData` is marked as optional field, and passing any comment data produces error:
* https://github.com/expo/expo/blob/master/docs/components/plugins/api/APIDataTypes.ts

Also I don't think that marking `comment` as required and changing its type to `CommentData | undefined` is a good solution in this situation, so maybe we can leave the code as is?

ide replied:
Ok. I still think it is better to lift this logic and write `comment ? getTagData(x, comment) : null` instead of having `getTagData` abstract this away.

Simek replied:
This might be a good idea, I will look at that and try to prepare separate PR with refactor in the future.

Simek committed:
apply review suggestions

Simek committed:
add basic CommentTextBlock tests

Simek commented:
As the comment structure is quite stable as this point, in the last commit I have added few test for the `CommentTextBlock` component. 

At the beginning I just want to only add test for the `@example` annotation, but finally I have also added few basic test and I have merged the `@example` test with other fields to reflect a bit more complicated use-case for the component.

Simek committed:
small tests tweaks
